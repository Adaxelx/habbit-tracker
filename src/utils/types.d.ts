export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: boolean;
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}
