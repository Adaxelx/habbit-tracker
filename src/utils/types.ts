export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
  handleDayChange: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: [
    {
      day: number;
      month: number;
      year: number;
      _id: string;
    },
  ];
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
  userId: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}

export type Color = { color?: string; id?: string };

export type NavProps = {
  header: string;
  moveDate: Function;
};

export interface DayCardProps {
  day: string;
  token: string | undefined;
}
