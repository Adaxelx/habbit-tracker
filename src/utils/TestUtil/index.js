export { default } from './TestUtil.ts';
