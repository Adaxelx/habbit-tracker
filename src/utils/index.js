export { default as TestUtil } from './TestUtil';
export { default as RenderWithRouter } from './RenderWithRouter';
export * from './validators';
export * from './types.d';
