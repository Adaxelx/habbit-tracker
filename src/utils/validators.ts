export const createRequiredObject = (name: string) => ({
  required: `Field ${name} is required.`,
});

export const validatePattern = (pattern: RegExp) => ({
  pattern: {
    value: pattern,
    message: 'Incorrect characters.',
  },
});

export const emailValidation = validatePattern(
  // eslint-disable-next-line no-control-regex
  /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/,
);

export const createRestrictedLengthObject = (name: string) => ({
  ...createRequiredObject(name),
  minLength: {
    value: 6,
    message: `Field ${name} must have at least 6 characters.`,
  },
  maxLength: {
    value: 30,
    message: `Field ${name} must have maximum 30 characters.`,
  },
});

const timeValidFuntion = (timeLess: string | undefined, time: string | undefined) => {
  const message = 'Your time is not in boundaries.';
  if (timeLess && time) {
    const [hours, minutes] = time.split(':');
    const [hoursMin, minutesMin] = timeLess.split(':');
    if (hoursMin < hours) {
      return true;
    }
    if (hoursMin === hours) {
      if (minutesMin < minutes) {
        return true;
      }
      return message;
    }
    return message;
  }
  return message;
};

export const createTimeValidation = (min: string | undefined, max: string | undefined) => ({
  validate: {
    min: (value: string) => timeValidFuntion(min, value),
    max: (value: string) => timeValidFuntion(value, max),
  },
});
