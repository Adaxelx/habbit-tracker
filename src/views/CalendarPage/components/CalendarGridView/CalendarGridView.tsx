import React, { useState } from 'react';
import { Button } from 'components';
import { StyledButtonWrapper } from 'views/CalendarPage/CalendarPage.css';
import { GridViewProps } from 'utils';
import { CalendarGrid, HabbitForm, LabelForm, LabelList } from '..';

const CalendarGridView = ({
  days,
  actualMonth,
  moveDate,
  actualYear,
  handleChangeView,
  labels,
}: GridViewProps) => {
  const [openHabbitForm, setOpenHabbitForm] = useState(false);
  const [openLabelForm, setOpenLabelForm] = useState(false);
  const [openLabelList, setOpenLabelList] = useState(false);

  return (
    <>
      <CalendarGrid
        days={days}
        month={actualMonth}
        moveDate={moveDate}
        year={actualYear}
        handleDayChange={handleChangeView}
      />
      <StyledButtonWrapper>
        <Button
          size="s"
          noMaxWidth
          mt="16px"
          data-testid="addh"
          disabled={!navigator.onLine}
          onClick={() => setOpenHabbitForm(true)}
        >
          Add habbit
        </Button>
        <Button
          size="s"
          noMaxWidth
          mt="16px"
          disabled={!navigator.onLine}
          mx="16px"
          data-testid="addl"
          onClick={() => setOpenLabelForm(true)}
        >
          Add label
        </Button>
        <Button
          size="s"
          noMaxWidth
          mt="16px"
          data-testid="labell"
          onClick={() => setOpenLabelList(true)}
        >
          Label list
        </Button>
      </StyledButtonWrapper>
      <HabbitForm
        open={openHabbitForm}
        handleClose={() => setOpenHabbitForm(false)}
        labels={labels}
      />
      <LabelForm open={openLabelForm} handleClose={() => setOpenLabelForm(false)} />
      <LabelList open={openLabelList} handleClose={() => setOpenLabelList(false)} labels={labels} />
    </>
  );
};

export default CalendarGridView;
