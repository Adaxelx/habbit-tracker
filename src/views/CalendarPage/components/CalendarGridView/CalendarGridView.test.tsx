import React from 'react';
import 'jest-styled-components';
import { TestUtil, GridViewProps } from 'utils';
import CalendarGridView from './CalendarGridView';

describe('CalendarGridView', () => {
  let util: TestUtil;

  let props: GridViewProps;

  beforeEach(() => {
    jest.clearAllMocks();

    props = {
      days: [{ day: 1, events: [], id: 1, active: true, date: new Date('2021-06-06') }],
      actualMonth: 1,
      moveDate: jest.fn(),
      actualYear: 1,
      handleChangeView: jest.fn(),
      labels: [{ _id: '123', userId: '123', title: 'label', color: '#676767' }],
    };
    util = new TestUtil(<CalendarGridView {...props} />);
  });

  it('should match snapshot', () => {
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('should close and open event form', () => {
    const header = 'Add habbit';
    util.click('addh');
    util.get(`${header}-header`);
    util.click(`${header}-close`);
  });

  it('should close and open label form', () => {
    const header = 'Add label';
    util.click('addl');
    util.get(`${header}-header`);
    util.click(`${header}-close`);
  });

  it('should close and open label list', () => {
    const header = 'Label list';
    util.click('labell');
    util.get(`${header}-header`);
    util.click(`${header}-close`);
  });
});
