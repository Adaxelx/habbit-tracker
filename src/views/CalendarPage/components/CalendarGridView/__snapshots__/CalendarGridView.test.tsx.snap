// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarGridView should match snapshot 1`] = `
<DocumentFragment>
  .c7 {
  border: none;
  width: 40px;
  height: 40px;
  position: relative;
  overflow: hidden;
  background-color: #FFFFFF;
}

.c7:hover {
  cursor: pointer;
}

.c8 {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 100;
}

.c0 {
  margin-top: 5rem;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(7,minmax(0,1fr));
  grid-template-rows: auto;
  max-width: 100%;
  border: 1px solid #707070;
  border-top: none;
  background-color: #fff;
}

.c5 {
  background-color: #EDEFFF;
  border-top: 1px solid #707070;
  border-bottom: none;
  justify-items: center;
}

.c6 {
  width: 40px;
  padding: 0.75rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0.75rem 0;
}

.c2 {
  background-color: transparent;
  border: none;
  font-size: 1.5rem;
}

.c3 {
  width: 200px;
  text-align: center;
  line-height: 100%;
  margin: 0;
}

@media (min-width:360px) {
  .c7 {
    width: 50px;
    height: 50px;
  }
}

@media (min-width:768px) {
  .c7 {
    width: 100px;
    height: 100px;
  }
}

@media (min-width:768px) {
  .c3 {
    width: 400px;
    font-size: 2rem;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
        data-testid="moveLeft-main"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-left fa-w-6 "
          data-icon="caret-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
            fill="currentColor"
          />
        </svg>
      </button>
      <h3
        class="c3"
        data-testid="dateCalendar-main"
      >
        February 1
      </h3>
      <button
        class="c2"
        data-testid="moveRight-main"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-right fa-w-6 "
          data-icon="caret-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="c4 c5"
    >
      <div
        class="c6"
      >
        Mon
      </div>
      <div
        class="c6"
      >
        Tue
      </div>
      <div
        class="c6"
      >
        Wed
      </div>
      <div
        class="c6"
      >
        Thu
      </div>
      <div
        class="c6"
      >
        Fri
      </div>
      <div
        class="c6"
      >
        Sat
      </div>
      <div
        class="c6"
      >
        Sun
      </div>
    </div>
    <div
      class="c4"
    >
      <button
        class="c7"
        data-testid="1-tileDay"
      >
        <p
          class="c8"
          data-testid="day"
        >
          1
        </p>
      </button>
    </div>
  </div>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c1 {
  color: #fff;
  background-color: #FFC136;
  border: 1px solid #707070;
  padding: 0.75rem;
  font-size: 1rem;
  min-width: none;
  display: inline-block;
  text-align: center;
  margin-top: 16px;
  margin-left: 0;
  margin-right: 0;
}

.c1:hover {
  cursor: pointer;
}

.c1:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

.c2 {
  color: #fff;
  background-color: #FFC136;
  border: 1px solid #707070;
  padding: 0.75rem;
  font-size: 1rem;
  min-width: none;
  display: inline-block;
  text-align: center;
  margin-top: 16px;
  margin-left: 16px;
  margin-right: 16px;
}

.c2:hover {
  cursor: pointer;
}

.c2:disabled {
  -webkit-filter: grayscale(1);
  filter: grayscale(1);
}

@media (min-width:360px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="addh"
    >
      Add habbit
    </button>
    <button
      class="c2"
      data-testid="addl"
    >
      Add label
    </button>
    <button
      class="c1"
      data-testid="labell"
    >
      Label list
    </button>
  </div>
</DocumentFragment>
`;
